drivers/dma/dma_mcux_smartdma.c

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/drivers/dma/dma_mcux_smartdma.h>
+
+#include <soc.h>
+#include <fsl_smartdma.h>
+#include <fsl_inputmux.h>
+#include <fsl_power.h>
+
+#define DT_DRV_COMPAT nxp_smartdma
+
+LOG_MODULE_REGISTER(dma_mcux_smartdma, CONFIG_DMA_LOG_LEVEL);
+
+/* SMARTDMA peripheral registers, taken from MCUX driver implementation*/
+struct smartdma_periph {
+	volatile uint32_t BOOT;
+	volatile uint32_t CTRL;
+	volatile uint32_t PC;
+	volatile uint32_t SP;
+	volatile uint32_t BREAK_ADDR;
+	volatile uint32_t BREAK_VECT;
+	volatile uint32_t EMER_VECT;
+	volatile uint32_t EMER_SEL;
+	volatile uint32_t ARM2SMARTDMA;
+	volatile uint32_t SMARTDMA2ARM;
+	volatile uint32_t PENDTRAP;
+};
+
+struct dma_mcux_smartdma_config {
+	struct smartdma_periph *base;
+	void (*irq_config_func)(const struct device *dev);
+
+	void (**smartdma_progs)(void);
+};
+
+struct dma_mcux_smartdma_data {
+	uint32_t smartdma_stack[32]; /* Stack for SMARTDMA */
+	/* Installed DMA callback and user data */
+	dma_callback_t callback;
+	void *user_data;
+};
+
+/* Seems to be written to smartDMA control register when it is configured */
+#define SMARTDMA_MAGIC 0xC0DE0000U
+/* These bits are set when the SMARTDMA boots, cleared to reset it */
+#define SMARTDMA_BOOT 0x11
+
+/* Configure a channel */
+static int dma_mcux_smartdma_configure(const struct device *dev,
+				uint32_t channel, struct dma_config *config)
+{
+	const struct dma_mcux_smartdma_config *dev_config = dev->config;
+	struct dma_mcux_smartdma_data *data = dev->data;
+	uint32_t prog_idx;
+	bool swap_pixels = false;
+
+	/* SMARTDMA does not have channels, so we only support channel 0. */
+	if (channel != 0) {
+		return -ENOTSUP;
+	}
+
+	data->callback = config->dma_callback;
+	data->user_data = config->user_data;
+
+	/* Reset smartDMA */
+	SMARTDMA_Reset();
+
+	/*
+	 * The dma_slot parameter is used to determine which SMARTDMA program
+	 * to run. First, convert the Zephyr define to a HAL enum.
```

Comment:
why not just #define the zephyr included macros as the hal enums instead of arbitrary constants, then you dont need to decode it at runtime or have this big switch-case